src/v/kafka/server/handlers/txn_offset_commit.cc

```diff
@@ -22,14 +22,137 @@
 
 namespace kafka {
 
+struct txn_offset_commit_ctx {
+    request_context rctx;
+    txn_offset_commit_request request;
+    ss::smp_service_group ssg;
+
+    // topic partitions found not to existent prior to processing. responses for
+    // these are patched back into the final response after processing.
+    absl::flat_hash_map<
+      model::topic,
+      std::vector<txn_offset_commit_response_partition>>
+      nonexistent_tps;
+
+    txn_offset_commit_ctx(
+      request_context&& rctx,
+      txn_offset_commit_request&& request,
+      ss::smp_service_group ssg)
+      : rctx(std::move(rctx))
+      , request(std::move(request))
+      , ssg(ssg) {}
+};
+
 template<>
 ss::future<response_ptr> txn_offset_commit_handler::handle(
-  request_context ctx, [[maybe_unused]] ss::smp_service_group g) {
-    return ss::do_with(std::move(ctx), [](request_context& ctx) {
-        txn_offset_commit_request request;
-        request.decode(ctx.reader(), ctx.header().version);
-        return ss::make_exception_future<response_ptr>(std::runtime_error(
-          fmt::format("Unsupported API {}", ctx.header().key)));
+  request_context ctx, ss::smp_service_group ssg) {
+    txn_offset_commit_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+
+    klog.trace("Handling request {}", request);
+
+    txn_offset_commit_ctx octx(std::move(ctx), std::move(request), ssg);
+
+    /*
+     * offset commit will operate normally on topic-partitions in the request
+     * that exist, while returning partial errors for those that do not exist.
+     * in order to deal with this we filter out nonexistent topic-partitions,
+     * and pass those that exist on to the group membership layer.
+     *
+     * TODO: the filtering is expensive for large requests. there are two things
+     * that can be done to speed this up. first, the metadata cache should
+     * provide an interface for efficiently searching for topic-partitions.
+     * second, the code generator should be extended to allow the generated
+     * structures to contain extra fields. in this case, we could introduce a
+     * flag to mark topic-partitions to be ignored by the group membership
+     * subsystem.
+     */
+    for (auto it = octx.request.data.topics.begin();
```

Comment:
I think we should break this up into 2 methods. 

1. filter
2. `do_handle` that does the `do_with()` biz below.

Reply:
I've copied it from `offset_commit.cc` let's change both places at once in a different PR